ArduCopter/ekf_check.cpp

```diff
@@ -160,6 +136,17 @@ void Copter::failsafe_ekf_event()
     failsafe.ekf = true;
     Log_Write_Error(ERROR_SUBSYSTEM_FAILSAFE_EKFINAV, ERROR_CODE_FAILSAFE_OCCURRED);
 
+    // does this mode require position?
+    if (!copter.flightmode->requires_GPS() && (g.fs_ekf_action != FS_EKF_ACTION_LAND_EVEN_STABILIZE)) {
```

Comment:
This shouldn't be here:

- GCS message and logging of failsafe will already have happened
- when we change mode (to other than Acro or Stabilize) the failsafe won't be triggered

Reply:
I think this behaviour is more consistent with our other failsafes.  So we actually go into the failsafe whether or not the flight mode require a good position.  The response to the failsafe is what's different depending upon the flight mode.  Then we do not allow the vehicle to go into modes requiring a position if the ekf failsafe has triggered.

I don't think we need to separate the EKF failsafe into two parts because we only care whether the vehicle has a good position and velocity estimate.  Having a bad compass, on its own, doesn't mean we need to trigger a failsafe.  If we can control the vehicle correctly, the vehicle can carry on flying.  There is reporting of the bad compass through the EKF variances message.

What's very bad about the current structure of the ekf failsafe in master is that it has two paths when really only one is required.

Reply:
> I think this behaviour is more consistent with our other failsafes. So we actually go into the failsafe whether or not the flight mode require a good position. The response to the failsafe is what's different depending upon the flight mode. Then we do not allow the vehicle to go into modes requiring a position if the ekf failsafe has triggered.

OK, but this means that if the failsafe was because of compass or velocity we will still enter other modes because they just check that position is OK.

> I don't think we need to separate the EKF failsafe into two parts because we only care whether the vehicle has a good position and velocity estimate. Having a bad compass, on its own, doesn't mean we need to trigger a failsafe. If we can control the vehicle correctly, the vehicle can carry on flying. There is reporting of the bad compass through the EKF variances message.

I'm confused then. Why do we trigger the failsafe based on mag variance? If we don't care about it then we don't need to check it.

> What's very bad about the current structure of the ekf failsafe in master is that it has two paths when really only one is required.

There aren't exactly two paths, there are two ways to trigger the failsafe, either because variances go up or because we don't know our position.
I would say that what is bad and unnecessary is that we are checking if we know our position, but also separately checking the position variance.

My comments here come from the long discussion in #6392, especially https://github.com/ArduPilot/ardupilot/pull/6392#issuecomment-329805648 and from https://github.com/ArduPilot/ardupilot/pull/6392#issuecomment-330530015 to the end.
@peterbarker and I have spent quite a bit of time in that PR trying to "fix" this failsafe, so I would like to not go around in circles changing stuff.

I think that we should define at a high level what we want. From your comments and my view it would be "provide a failsafe that executes an action when we need to know our position and AHRS can't provide it". That's not exactly an EKF failsafe though and has no compass involved.